File: python/medousa-sdk/src/medousa/http.py
from __future__ import annotations

from typing import Any, TypeVar

from pydantic import BaseModel

from medousa._decode import decode
from medousa.client import MedousaClient
from medousa.transport import path_with_query

T = TypeVar("T", bound=BaseModel)


class HttpApi:
    def __init__(self, client: MedousaClient) -> None:
        self._client = client

    async def get(self, path: str, *, model: type[T]) -> T:
        value = await self._client.transport.get_json(self._client.base_url, path)
        return decode(model, value)

    async def get_query(
        self,
        path: str,
        query: list[tuple[str, str]],
        *,
        model: type[T],
    ) -> T:
        return await self.get(path_with_query(path, query), model=model)

    async def post(self, path: str, body: Any, *, model: type[T]) -> T:
        value = await self._client.transport.post_json(
            self._client.base_url,
            path,
            _serialize(body),
        )
        return decode(model, value)

    async def post_empty(self, path: str, *, model: type[T]) -> T:
        value = await self._client.transport.post_empty_json(self._client.base_url, path)
        return decode(model, value)

    async def put(self, path: str, body: Any, *, model: type[T]) -> T:
        value = await self._client.transport.put_json(
            self._client.base_url,
            path,
            _serialize(body),
        )
        return decode(model, value)

    async def patch(self, path: str, body: Any, *, model: type[T]) -> T:
        value = await self._client.transport.patch_json(
            self._client.base_url,
            path,
            _serialize(body),
        )
        return decode(model, value)

    async def delete(self, path: str, *, model: type[T]) -> T:
        value = await self._client.transport.delete_json(self._client.base_url, path)
        return decode(model, value)

    async def post_raw(self, path: str, body: Any) -> Any:
        return await self._client.transport.post_json(
            self._client.base_url,
            path,
            _serialize(body),
        )

    async def delete_raw(self, path: str) -> Any:
        return await self._client.transport.delete_json(self._client.base_url, path)


def _serialize(body: Any) -> Any:
    if isinstance(body, BaseModel):
        return body.model_dump(mode="json", exclude_none=True)
    return body

File: crates/medousa-sdk/src/http.rs
#[cfg(feature = "async")]
use serde::{Serialize, de::DeserializeOwned};

#[cfg(feature = "async")]
use crate::client::MedousaClient;
#[cfg(feature = "async")]
use crate::transport::{decode, path_with_query};

#[cfg(feature = "async")]
pub struct HttpApi<'a> {
    pub(crate) client: &'a MedousaClient,
}

#[cfg(feature = "async")]
impl HttpApi<'_> {
    pub async fn get<T: DeserializeOwned>(&self, path: &str) -> Result<T, crate::SdkError> {
        let value = self
            .client
            .transport()
            .get_json(self.client.base_url(), path)
            .await?;
        decode(value).await
    }

    pub async fn get_query<T: DeserializeOwned>(
        &self,
        path: &str,
        query: &[(&str, String)],
    ) -> Result<T, crate::SdkError> {
        let path = path_with_query(path, query);
        self.get(&path).await
    }

    pub async fn post<T: DeserializeOwned, B: Serialize>(
        &self,
        path: &str,
        body: &B,
    ) -> Result<T, crate::SdkError> {
        let body = serde_json::to_value(body).map_err(|e| crate::SdkError::Serde(e.to_string()))?;
        let value = self
            .client
            .transport()
            .post_json(self.client.base_url(), path, body)
            .await?;
        decode(value).await
    }

    pub async fn post_empty<T: DeserializeOwned>(&self, path: &str) -> Result<T, crate::SdkError> {
        let value = self
            .client
            .transport()
            .post_empty_json(self.client.base_url(), path)
            .await?;
        decode(value).await
    }

    pub async fn put<T: DeserializeOwned, B: Serialize>(
        &self,
        path: &str,
        body: &B,
    ) -> Result<T, crate::SdkError> {
        let body = serde_json::to_value(body).map_err(|e| crate::SdkError::Serde(e.to_string()))?;
        let value = self
            .client
            .transport()
            .put_json(self.client.base_url(), path, body)
            .await?;
        decode(value).await
    }

    pub async fn patch<T: DeserializeOwned, B: Serialize>(
        &self,
        path: &str,
        body: &B,
    ) -> Result<T, crate::SdkError> {
        let body = serde_json::to_value(body).map_err(|e| crate::SdkError::Serde(e.to_string()))?;
        let value = self
            .client
            .transport()
            .patch_json(self.client.base_url(), path, body)
            .await?;
        decode(value).await
    }

    pub async fn delete<T: DeserializeOwned>(&self, path: &str) -> Result<T, crate::SdkError> {
        let value = self
            .client
            .transport()
            .delete_json(self.client.base_url(), path)
            .await?;
        decode(value).await
    }
}
